### ISSUE TITLE
creating temp table will generate new scheamas

### ISSUE THREAD

**Issue by lij55** [greenplum-db/gpdb#4061]
### Greenplum version or build
GPDB 5.2 (and 5.3)
### OS version and uname -a
Linux 07fd26f80ae4 3.10.0-229.el7.x86_64 #1 SMP Fri Mar 6 11:36:42 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

### Expected behavior
Run following query for several times first: `psql -c   "create temp table a();"` Then run `\dn` from psql command line tool.
There are a lot of pg_toast_temp_SESSIONID schema in GPDB5 but not in GPDB4.
```
test=# \dn
       List of schemas
        Name        |  Owner
--------------------+---------
 gp_toolkit         | gpadmin
 information_schema | gpadmin
 pg_aoseg           | gpadmin
 pg_bitmapindex     | gpadmin
 pg_catalog         | gpadmin
 pg_toast           | gpadmin
 pg_toast_temp_1    | gpadmin
 pg_toast_temp_11   | gpadmin
 pg_toast_temp_12   | gpadmin
 pg_toast_temp_13   | gpadmin
 pg_toast_temp_14   | gpadmin
 pg_toast_temp_15   | gpadmin
 pg_toast_temp_16   | gpadmin
 public             | gpadmin
(14 rows)

test=# select version();
                                                                                   version

----------------------------------------------------------------------------------------------------
-------------------------------------------------------------------------
 PostgreSQL 8.3.23 (Greenplum Database 5.2.0 build dev) on x86_64-pc-linux-gnu, compiled by GCC gcc
(GCC) 4.8.5 20150623 (Red Hat 4.8.5-16) compiled on Nov 23 2017 16:33:05
(1 row)
```
GPDB 4.3.18:
```
psql (8.2.15)
Type "help" for help.

test=# \dn
       List of schemas
        Name        |  Owner
--------------------+---------
 gp_toolkit         | gpadmin
 information_schema | gpadmin
 pg_aoseg           | gpadmin
 pg_bitmapindex     | gpadmin
 pg_catalog         | gpadmin
 pg_toast           | gpadmin
 public             | gpadmin
(7 rows)

test=# select version();
                                                                        version

----------------------------------------------------------------------------------------------------
---------------------------------------------------
 PostgreSQL 8.2.15 (Greenplum Database 4.3.18.0 build 1) on x86_64-unknown-linux-gnu, compiled by GC
C gcc (GCC) 4.4.2 compiled on Nov 22 2017 18:54:31
(1 row)
```

**Comment by hlinnaka** [comment]
Yeah, that's expected. Since PostgreSQL 8.3, commit 82eed4dba2, toast tables for temp tables are put in those pg_toast_temp_XX schemas, similar to pg_temp_XX. This got fixed in 9.1, with commit e43fb604d6.

**Comment by lij55** [comment]
Hi @hlinnaka , the fix https://github.com/greenplum-db/gpdb/commit/e43fb604d6db229d70d3101aa53348cc16a5473a looks like just hide the schma simply, is it correct? 
Does it make sense to keep so many unusable empty temp schema in GPDB 5?

**Comment by water32** [comment]
These schema named as pg_toast_temp_sessionid growing and cannot be normal deleted , pg_namespace's record will growing without limit, this is an unacceptable feature, our PS team worry that this will cause stability problems, we may cannot continue to promote the using of 5.x, to force delete these schemas,  we must manually configure allow_system_table_mods= DDL | ALL in postgresql.conf, this configuration is extremely dangerous, if this is a bug introduced in PG8.3, it should be fixed in 5.X, rather than after upgrading to PG9.1.

**Comment by hlinnaka** [comment]
On 08/12/17 06:26, water32 wrote:
> These schema named as pg_toast_temp_sessionid growing and cannot be
> normal deleted , pg_namespace's record will growing without limit,

Ah, now I understand what the problem is. I thought the problem was just 
that they're shown in \dn output, I didn't realize that they're not 
being cleaned up.

The pg_toast_temp_* namespaces work the same as the pg_temp_* namespaces 
we've always had. For each session, you get one pg_temp_<sessionid> 
namespace, and one pg_temp_<sessionid> namespace. The difference is that 
there's code that runs at backend exist, that drops the 
"pg_temp_<sessionid>" schema, but it doesn't do the same for the 
"pg_toast_temp_<sessionid>" schema. I'll go and fix that.

In PostgreSQL, we use the backend ID instead of the GPDB-specific 
session ID. That limits the growth, as the backend IDs go only from 1 to 
max_connections. In fact, in the upstream, we *don't* remove the temp 
schemas at backend exit. We only remove their contents. That saves the 
overhead of re-creating the schema, for the next backend with the same 
ID. In the long run, it would be good to do that in GPDB too. Currently, 
if the server is goes down abruptly, without a clean shutdown, the temp 
namespaces will be left behind.

> this is an unacceptable feature, our PS team worry that this will
> cause stability problems, we may cannot continue to promote the using
> of 5.x, to force delete these schemas,  we must manually configure
> allow_system_table_mods= DDL | ALL in postgresql.conf, this
> configuration is extremely dangerous, if this is a bug introduced in
> PG8.3, it should be fixed in 5.X, rather than after upgrading to
> PG9.1.

Chill out, we'll fix it :-).

So the immediate fix is to drop the temp toast namespace, at the same 
places we drop the temp namespace. I'll go and open a PR for that.

In the long run, we should consider naming the temp namespaces 
differently, perhaps using the QD backend's backendid rather than the 
session id, to avoid bloating the catalogs if a backend exists abnormally.

- Heikki

**Comment by hlinnaka** [comment]
Fixed by PR #4092. Thanks for the report, and sorry for my earlier confusion on what the problem was!

**Comment by lij55** [comment]
Thanks, Heikki. I will try to explain the key problem clearly enough next time.
